Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -0,0 +1,572 @@
+                                                    { CFSTR("EXTRABOLD"), DWRITE_FONT_WEIGHT_EXTRA_BOLD },
+                                                    { CFSTR("ULTRABOLD"), DWRITE_FONT_WEIGHT_ULTRA_BOLD },
+                                                    { CFSTR("BLACK"), DWRITE_FONT_WEIGHT_BLACK },
+                                                    { CFSTR("HEAVY"), DWRITE_FONT_WEIGHT_HEAVY },
+                                                    { CFSTR("EXTRABLACK"), DWRITE_FONT_WEIGHT_EXTRA_BLACK },
+                                                    { CFSTR("ULTRABLACK"), DWRITE_FONT_WEIGHT_ULTRA_BLACK } };
+
+/**
+ * Helper that parses a font name, and returns appropriate weight, stretch, style, and family name values
+ */
+void _InitDWriteFontPropertiesFromName(
+    CFStringRef fontName, DWRITE_FONT_WEIGHT* weight, DWRITE_FONT_STRETCH* stretch, DWRITE_FONT_STYLE* style, CFStringRef* familyName) {
+    // Set some defaults for when weight/stretch/style are not mentioned in the name
+    // Since this is a file-private helper, assume the pointers are safe to dereference.
+    *weight = DWRITE_FONT_WEIGHT_NORMAL;
+    *stretch = DWRITE_FONT_STRETCH_NORMAL;
+    *style = DWRITE_FONT_STYLE_NORMAL;
+
+    *familyName = _DWriteGetFamilyNameForFontName(fontName);
+
+    // Relationship of family name -> font name not always consistent
+    // Usually, properties are added to the end (eg: Arial -> Arial Narrow Bold)
+    // However, this is not always the case (eg: Eras ITC -> Eras Bold ITC)
+    // In addition, some fonts with properties are occasionally placed into their own family (eg: Segoe WP SemiLight -> Segoe WP SemiLight)
+    // Try to be more prudent about these edge cases, by looking only at the difference between the font name and family name
+    CFArrayRef fontNameTokens = CFStringCreateArrayBySeparatingStrings(kCFAllocatorDefault, fontName, CFSTR(" "));
+    CFAutorelease(fontNameTokens);
+    CFMutableSetRef propertyTokens = CFSetCreateMutable(kCFAllocatorDefault, 0, &kCFTypeSetCallBacks);
+    CFAutorelease(propertyTokens);
+
+    for (size_t i = 0; i < CFArrayGetCount(fontNameTokens); ++i) {
+        CFSetAddValue(propertyTokens, CFArrayGetValueAtIndex(fontNameTokens, i));
+    }
+
+    if (*familyName) {
+        CFArrayRef familyNameTokens = CFStringCreateArrayBySeparatingStrings(kCFAllocatorDefault, *familyName, CFSTR(" "));
+        CFAutorelease(familyNameTokens);
+        for (size_t i = 0; i < CFArrayGetCount(familyNameTokens); ++i) {
+            CFSetRemoveValue(propertyTokens, CFArrayGetValueAtIndex(familyNameTokens, i));
+        }
+    }
+
+    // Store weight, stretch, and style in a single struct to be passed into a CFSetApplierFunction's context
+    struct PropertyContext {
+        DWRITE_FONT_WEIGHT* weight;
+        DWRITE_FONT_STRETCH* stretch;
+        DWRITE_FONT_STYLE* style;
+    };
+
+    struct PropertyContext propertyContext = { weight, stretch, style };
```

Comment:
usually points to a "curious" design decision in the first place. In this case you have a void return function with three out parameters and then a bunch of weird glue code to make that work. How about defining the struct outside the function and using it as a return type? Then you don't need to worry about the modification contracts with the caller's data.

---
In reply to: [84400714](https://github.com/Microsoft/WinObjC/pull/1145#discussion_r84400714) [](ancestors = 84400714,83338922)